### PR TITLE
Copy loss and metric to prevent side effect

### PR DESCRIPTION
**Pull request by drauh**
This is a possible solution for #16356, a possible error that could be raised when saving weights. This happens because the inputted dict to the compile method is modified by a side effect, modifying also the keras.Model.loss attribute.

### CODE REVIEW

File: keras/engine/compile_utils.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def __init__(self, losses, loss_weights=None, output_names=None):
     self._user_losses = losses
     self._user_loss_weights = loss_weights
 
-    self._losses = losses
+    self._losses = copy.copy(losses)
```
Comment (fchollet) [inline keras/engine/compile_utils.py:110]: As far as I can tell (cf. the `compile()` method) the `losses` argument is *not* guaranteed to be a list/dict and thus is not safe to copy.

File: keras/engine/compile_utils.py [hunk 2]
```diff
@@ -309,7 +309,7 @@ def __init__(self, metrics=None, weighted_metrics=None, output_names=None,
     self._user_metrics = metrics
     self._user_weighted_metrics = weighted_metrics
 
-    self._metrics = metrics
+    self._metrics = copy.copy(metrics)
```
Comment (fchollet) [inline keras/engine/compile_utils.py:312]: The same is applicable to metrics.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR.

> This happens because the inputted dict to the compile method is modified by a side effect

Do you know where this happens, specifically?

**Comment by drauh** [comment]
It's happening in [`map_missing_dict_keys`](https://github.com/keras-team/keras/blob/3c63251dde6ea9df2787952a9687fbf14366e5de/keras/engine/compile_utils.py#L57). Here after the check, it's guaranteed to be a dict so I made the copy there. The test I wrote to check this issue is passing also with this solution.

**Review by fchollet — APPROVED** [review]
That looks good! Thank you.
